src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,505 @@
+
+                while (visitStack.Count > 0)
+                {
+                    var current = visitStack.Pop();
+                    var setToAddTo = capturesVariable.Contains(current) ? capturesVariable : capturesThis;
+                    foreach (var capturesCurrent in methodGraph[current])
+                    {
+                        if (setToAddTo.Add(capturesCurrent))
+                        {
+                            visitStack.Push(capturesCurrent);
+                        }
+                    }
+                }
+
+                VisitClosures(_scopeTree, (scope, closure) =>
+                {
+                    if (!capturesVariable.Contains(closure.OriginalMethodSymbol))
+                    {
+                        closure.CapturedVariables.Clear();
+                    }
+                    if (capturesThis.Contains(closure.OriginalMethodSymbol))
+                    {
+                        closure.CapturedVariables.Add(_topLevelMethod.ThisParameter);
+                    }
+                });
+            }
+
+            /// <summary>
+            /// Visit all closures in all nested scopes and run the <paramref name="action"/>.
+            /// </summary>
+            private static void VisitClosures(Scope scope, Action<Scope, Closure> action)
+            {
+                foreach (var closure in scope.Closures)
+                {
+                    action(scope, closure);
+                }
+
+                foreach (var nested in scope.NestedScopes)
+                {
+                    VisitClosures(nested, action);
+                }
+            }
+
+            /// <summary>
+            /// Builds a tree of <see cref="Scope"/> nodes corresponding to a given method.
+            /// <see cref="Build(BoundNode, Analysis)"/> visits the bound tree and translates
+            /// information from the bound tree about variable scope, declared variables, and
+            /// variable captures into the resulting <see cref="Scope"/> tree.
+            /// </summary>
+            private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+            {
+                private Scope _currentScope;
+                /// <summary>
+                /// Null if we're not inside a closure scope, otherwise the nearest closure scope
+                /// </summary>
+                private Closure _currentClosure = null;
+                private bool _inExpressionTree = false;
+
+                /// <summary>
+                /// A mapping from all captured vars to the scope they were declared in. This
+                /// is used when recording captured variables as we must know what the lifetime
+                /// of a captured variable is to determine the lifetime of its capture environment.
+                /// </summary>
+                private readonly SmallDictionary<Symbol, Scope> _localToScope = new SmallDictionary<Symbol, Scope>();
```

Comment:
This is another place where having a new simple struct type that wrapped `Symbol` would be advantageous. It would allow you to add the `Scope` as a field. This avoids the need for arbitrary dictionaries like this (which this PR is overall trying to replace).